File: services/api/src/restaurant_os_api/modules/operations/domain/entities/cash_drawer.py
"""CashDrawer entity. The Blueprint's "daily cash-up automatically
reconciled against POS sales" story reads a closed drawer against the
sum of its shift's cash Payments -- a query the use case layer performs
(``CloseCashDrawerUseCase``), not a separate reconciliation entity."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from decimal import Decimal
from enum import StrEnum

from restaurant_os_api.modules.operations.domain.exceptions import (
    InvalidCashDrawerStatusTransitionError,
)


class CashDrawerStatus(StrEnum):
    OPEN = "open"
    CLOSED = "closed"


@dataclass(slots=True)
class CashDrawer:
    id: str
    tenant_id: str
    branch_id: str
    status: CashDrawerStatus
    opening_float_amount: Decimal
    opened_at: datetime
    created_at: datetime
    terminal_id: str | None = None
    closing_counted_amount: Decimal | None = None
    closed_at: datetime | None = None

    def close(self, *, closing_counted_amount: Decimal, closed_at: datetime) -> None:
        if self.status != CashDrawerStatus.OPEN:
            raise InvalidCashDrawerStatusTransitionError(
                self.id, self.status.value, CashDrawerStatus.CLOSED.value
            )
        self.status = CashDrawerStatus.CLOSED
        self.closing_counted_amount = closing_counted_amount
        self.closed_at = closed_at

File: apps/admin-web/src/types/cash-drawer.ts
/**
 * Mirrors modules/operations/presentation/api/v1/cash_drawer_router.py's
 * CashDrawerResponseSchema / Open·Close RequestSchemas and the domain's
 * CashDrawerStatus StrEnum. Field names are camelCase on the wire; money
 * fields are decimal strings. ``expectedCashAmount``/``varianceAmount``
 * are computed at read time, only populated once the drawer is closed.
 */

export type CashDrawerStatus = "open" | "closed"

export interface CashDrawer {
  id: string
  tenantId: string
  branchId: string
  status: CashDrawerStatus
  openingFloatAmount: string
  openedAt: string
  createdAt: string
  terminalId: string | null
  closingCountedAmount: string | null
  closedAt: string | null
  expectedCashAmount: string | null
  varianceAmount: string | null
}

export interface OpenCashDrawerRequest {
  openingFloatAmount: string
  terminalId?: string | null
}

export interface CloseCashDrawerRequest {
  closingCountedAmount: string
}
